bids/layout/index.py

```diff
@@ -36,38 +40,92 @@ def _check_path_matches_patterns(path, patterns):
+        spec, as setting validate=True will lead files in supplementary folders
+        like derivatives/, code/, etc. to be ignored.
+    ignore : str or SRE_Pattern or list
+        Path(s) to exclude from indexing. Each path is either a string or a
+        SRE_Pattern object (i.e., compiled regular expression). If a string is
+        passed, it must be either an absolute path, or be relative to the BIDS
+        project root. If an SRE_Pattern is passed, the contained regular
+        expression will be matched against the full (absolute) path of all
+        files and directories. By default, indexing ignores all files in
+        'code/', 'stimuli/', 'sourcedata/', 'models/', and any hidden
+        files/dirs beginning with '.' at root level.
+    force_index : str or SRE_Pattern or list
+        Path(s) to forcibly index in the BIDSLayout, even if they would
+        otherwise fail validation. See the documentation for the ignore
+        argument for input format details. Note that paths in force_index takes
+        precedence over those in ignore (i.e., if a file matches both ignore
+        and force_index, it *will* be indexed).
+        Note: NEVER include 'derivatives' here; use the derivatives argument
+        (or :obj:`bids.layout.BIDSLayout.add_derivatives`) for that.
+    index_metadata : bool
+        If True, all metadata files are indexed. If False, metadata will not be
+        available (but indexing will be faster).
+    config_filename : str
+        Optional name of filename within directories
+        that contains configuration information.
+    **filters
+        keyword arguments passed to the .get() method of a
+        :obj:`bids.layout.BIDSLayout` object. These keyword arguments define
+        what files get selected for metadata indexing.
     """
 
-    def __init__(self, layout):
-
-        self.layout = layout
-        self.config = list(layout.config.values())
-        self.session = layout.session
-        self.validate = layout.validate
-        self.root = layout.root
-        self.config_filename = layout.config_filename
+    def __init__(self, validate=True, ignore=None, force_index=None,
+                 index_metadata=True, config_filename='layout_config.json',
+                 **filters):
+        self.validate = validate
+        self.ignore = ignore
+        self.force_index = force_index
+        self.index_metadata = index_metadata
+        self.config_filename = config_filename
+        self.filters = filters
         self.validator = BIDSValidator(index_associated=True)
-        # Create copies of list attributes we'll modify during indexing
-        self.include_patterns = list(layout.force_index)
-        self.exclude_patterns = list(layout.ignore)
+
+        # Layout-dependent attributes to be set in __call__()
+        self._layout = None
+        self._config = None
+        self._include_patterns = None
+        self._exclude_patterns = None
+
+    def __call__(self, layout):
```

Comment:
That was the original setup, and I've fiddled with it several times. I ended up going with this approach because it allows a user to pass in their own indexer if they want to—it can be any callable that takes a `BIDSLayout` as its only argument. I doubt anyone will want to, but you never know. Otherwise I don't really see a problem with this approach; it seems unlikely anyone will ever need to index from outside the `BIDSLayout`. It *would* make a difference if we wanted to allow incremental indexing, which I originally did (#522), but for reasons not worth getting into, that would be pretty difficult to do anyway without substantively rewriting the current indexing code. So I'm going to leave it as-is (but will update the docs to point out one can pass a callable indexer, which I forgot to do).